### DISCUSSION TITLE
SSD1306+lvgl sample fails to display

### DISCUSSION THREAD

**Discussion by zpm1066** [zephyrproject-rtos/zephyr#51023] [Q&A]
**Describe the bug**
I'm able to build the sample "zephyrproject/zephyr/samples/subsys/display/lvgl" for a nrf52840 board using an overlay. The program builds fine but a connected generic OLED SSD1306 128x64 display (also a ssd1306 Adafruit FeatherWing OLED - 128x64) fails to display anything. 
The sample code hasn't been modified.

Please also mention any information which could help others to understand
the problem you're facing:
 - What target platform are you using?
 nrf52840 board (particle_xenon)
 
 - What have you tried to diagnose or workaround this issue?
 In debug, it appears that the program crashes within lv_task_handler().
 
 - Is this a regression? 
 Unable to determine, as I cannot get the SSD1306 + lvgl to work with several Zephyr versions.

**To Reproduce**
1. Use the sample code "zephyrproject/zephyr/samples/subsys/display/lvgl" as is. 
2. Use the proj.conf and particle_xenon.overlay below.

**particle_xenon.overlay:**

```
/ {
	chosen {
		zephyr,display = &ssd1306;
	};
};


&pinctrl {
		i2c0_default: i2c0_default {
			group1 {
				psels = <NRF_PSEL(TWIM_SDA, 0, 26)>,
					<NRF_PSEL(TWIM_SCL, 0, 27)>;
			};
		};
	
		i2c0_sleep: i2c0_sleep {
			group1 {
				psels = <NRF_PSEL(TWIM_SDA, 0, 26)>,
					<NRF_PSEL(TWIM_SCL, 0, 27)>;
				low-power-enable;
			};
		};
};

&gpio0{
	status = "okay";
};


&gpio1{
	status = "okay";
};

&i2c0 {
	compatible = "nordic,nrf-twim";
	status = "okay";
	pinctrl-0 = <&i2c0_default>;
	pinctrl-1 = <&i2c0_sleep>;
    pinctrl-names = "default", "sleep";

	status = "okay";
	ssd1306: ssd1306@3c {
		compatible = "solomon,ssd1306fb";
		reg = <0x3c>;
		width = <128>;
		height = <64>;
		segment-offset = <0>;
		page-offset = <0>;
		display-offset = <0>;
		multiplex-ratio = <63>;
		segment-remap;
		com-invdir;
		prechargep = <0x22>;
	};
};

```

**proj.conf:**

```
CONFIG_MAIN_STACK_SIZE=2048

CONFIG_I2C=y
CONFIG_I2C_NRFX=y
CONFIG_LOG=y

CONFIG_DISPLAY=y
CONFIG_DISPLAY_LOG_LEVEL_ERR=y

CONFIG_SSD1306=y
CONFIG_SSD1306_DEFAULT=y

CONFIG_LVGL=y
CONFIG_LV_MEM_CUSTOM=y
CONFIG_LV_USE_LOG=y
CONFIG_LV_USE_LABEL=y
CONFIG_LV_USE_BTN=y
CONFIG_LV_USE_IMG=y
CONFIG_LV_FONT_MONTSERRAT_14=y

CONFIG_LV_Z_BITS_PER_PIXEL=1
CONFIG_LV_Z_MEM_POOL_NUMBER_BLOCKS=8
CONFIG_APP_LINK_WITH_LVGL=y

```

*Expected behavior**
I expected the lvgl sample to work with the ssd1306 display and show the "Hello world!" text in the middle of the display.

**Impact**
Unable to use SSD1306 + lvgl on several key projects.

**Logs and console output**
Program builds without any errors.

**Environment (please complete the following information):**
macOS Big Sur & Catalina
Zephyr RTOS v3.1 and 3.2

**Additional context**
The same SSD1306 display with exactly the same overlay & connections works fine with nrf52840 and other boards when using SSD1306 with cfb_custom_font & cfb samples.

**Comment by zpm1066** [comment]
@laurenmurphyx64, @henrikbrixandersen, @jfischer-no, @brgl 
I'm also going to create an account at lvgl.com to see if they can also investigate, as I don't see anything wrong with the Zephyr code. 

I have several types of displays (ST7789v, ili9341 - 1.14" to 2.8") that all work just fine with Zephyr + lvgl. SSD1306 works fine with Zephyr on it's own. There may be a SSD1306 binding issue with lvgl.

I'll be glad to assist with further debugging of SSD1306/lvgl. Thanks.

**Comment by jfischer-no** [comment]
> Zephyr RTOS v3.1 and 3.2

I cannot confirm, works fine for me using `west build -b nrf52840dk_nrf52840 samples/subsys/display/lvgl/ -- -DSHIELD=ssd1306_128x64`. Moving to discussion since it is likely configuration issue.

**Reply by zpm1066** [reply]
Hi @jfischer-no, Thanks. Did the text get displayed on the SSD1306?

For me everything builds without any errors but after flashing nothing gets displayed on the SSD1306.

Do you see anything wrong with the particle_xenon.overlay and proj.conf files? I didn’t use the overlay from the SHIELD but specified my own overlay, the same as what I do for the ST7789/ILI9341 displays.

I also included the ssd1306_128x64 SHIELD to check how it’s overlay is defined, matched what I have for my board.

any suggestions what I should check? Thank you.

**Reply by jfischer-no** [reply]
> Hi @jfischer-no, Thanks. Did the text get displayed on the SSD1306?

yes.

> For me everything builds without any errors but after flashing nothing gets displayed on the SSD1306.
> 
> Do you see anything wrong with the particle_xenon.overlay and proj.conf files? I didn’t use the overlay from the SHIELD but specified my own overlay, the same as what I do for the ST7789/ILI9341 displays.
> 
> I also included the ssd1306_128x64 SHIELD to check how it’s overlay is defined, matched what I have for my board.
> 
> any suggestions what I should check? Thank you.

Connect a SSD1306 display to I2C0 (TWI0 SDA/SCL) pins and try
west build -b particle_xenon samples/subsys/display/lvgl/ -- -DSHIELD=ssd1306_128x64.

**Reply by zpm1066** [reply]
Hi @jfischer-no, Thank you. The “ -DSHIELD=ssd1306_128x64” works!

However, I typically use a board overlay and cannot quite understand why using &i2c0 for ssd1306 in the above overlay doesn’t work. I have no problems with using the board overlays for the st7789/ili9341 displays instead of the SHIELD overlays.

Do you know why the overlay doesn’t work? It seems I’m missing some configuration.

Can you please try using a nrf52840dk_nrf52840.overlay with the ssd1306 device definition instead of the ssd1306_128x64 shield? Thank you.

**Reply by jfischer-no** [reply]
> Can you please try using a nrf52840dk_nrf52840.overlay with the ssd1306 device definition instead of the ssd1306_128x64 shield? Thank you.

It is unlikely that I will find the time, sorry.

**Reply by zpm1066** [reply]
Hi @jfischer-no , not a problem.

Do you see anything wrong with the overlay and proj.conf  I have defined for particle_xenon?

Any suggestion as to what I should check next? Thanks.

**Reply by henrikbrixandersen** [reply]
I suggest checking the differences between the generated zephyr.dts file when using the shield vs. using your overlay.

**Reply by zpm1066** [reply]
@jfischer-no , @henrikbrixandersen  - I've fixed the ` lvgl configuration` issues and now are able to use the board overlay instead of the `ssd1306_128x64 SHIELD` overlay. 

The fixed turned out be pretty simple. Earlier, I neglected to check the `Kconfig.defconfig` for the `ssd1306_128x64 SHIELD`. 

My `proj.conf` was missing entries to set color depth for ssd1306 to monochrome (ssd1306 default was 16). In addition, `LV_Z_VDB_SIZE` setting was missing, causing a crash in the lv_task_handler().

```
CONFIG_LV_USE_THEME_DEFAULT=y
CONFIG_LV_Z_VDB_SIZE=64
CONFIG_LV_COLOR_DEPTH_1=y
```

I also used the `LV_USE_THEME_DEFAULT`, so that in the code I could flip the display primary & secondary colours.

```
lv_disp_t * lv_disp = lv_disp_get_default();
lv_theme_default_init(lv_disp, lv_color_black(), lv_color_white(), 1, &lv_font_montserrat_14);
```

For the nRF52, nRF53, microbit v2, and nrf9160, I use the excellent nRF Connect for VSCode. Using the Kconfig GUI, I was able to note immediately the issues.

The built in SHIELD overlays are quite useful as well but generally I prefer to use the board overlays.

Thank you for your assistance.

**Comment by paulofpimenta** [comment]
Hi all , 

I am facing the same issue with a stm32F46ng_discovery. I will write down my configuration : 

Platform : STM32F746ng discovery

1. What have you tried to diagnose or workaround this issue?
Changing the proj file

2. Is this a regression?
No. The project works when I do  `west build /opt/toolchains/zephyr/samples/subsys/display/lvgl/ -b smt32f746g_disco -p`

3. To Reproduce

My main function as it is in the example : 

https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/subsys/display/lvgl/src/main.c

Define the proj file as : 

```
CONFIG_MAIN_STACK_SIZE=2048

CONFIG_I2C=y
CONFIG_LOG=y

CONFIG_DISPLAY=y
CONFIG_DISPLAY_LOG_LEVEL_ERR=y

CONFIG_LVGL=y
CONFIG_LV_USE_LOG=y
CONFIG_LV_USE_LABEL=y
CONFIG_LV_FONT_MONTSERRAT_14=y

CONFIG_LV_USE_THEME_DEFAULT=y
CONFIG_LV_Z_VDB_SIZE=64
CONFIG_LV_COLOR_DEPTH_1=y
```
And the `CMakeLists.txt`:

```
# Display : RK043FN48H-CT672B
cmake_minimum_required(VERSION 3.20.0)
set(BOARD_FLASH_RUNNER "openocd")

find_package(Zephyr REQUIRED HINTS $ENV{ZEPHYR_BASE})
project(lcd_hello)

FILE(GLOB app_sources src/*.c)
target_sources(app PRIVATE ${app_sources})
```

The KConfig:
```
source "Kconfig.zephyr"

mainmenu "LVGL Sample"

config RESET_COUNTER_SW0
	bool "Reset counter when sw0 is pressed"
	depends on $(dt_alias_enabled,sw0)
	select GPIO
	default y
```	 
The project builds and flashes without any issues run and the output is an empty black screen. Using Zephyr version 4.3.0 and SDK 0.17.4. Thank you in advance for any advices
